### PR TITLE
feat(pipeline): Add edge cluster pipeline reporting

### PR DESCRIPTION
**Pull request by littlejiancc**
#### What this PR does / why we need it:
 Add edge cluster pipeline reporting

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMTkwLDExNzRdLCJhc3NpZ25lZSI6WyIxMDAxMjYxIl19&id=300219&iterationID=1190&pId=306714&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |    Add edge cluster pipeline reporting          |
| 🇨🇳 中文    |     增加边缘集群流水线上报         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/edgereporter/provider.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+type provider struct {
+	bdl      *bundle.Bundle
+	dbClient *db.Client
+
+	Cfg          *config
+	Log          logs.Logger
+	LW           leaderworker.Interface
+	MySQL        mysqlxorm.Interface
+	edgeRegister edgepipeline_register.Interface
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply modules/pipeline/providers/edgereporter/provider.go:47]: done

File: modules/pipeline/providers/edgereporter/provider.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	"github.com/erda-project/erda/modules/pipeline/dbclient"
+	"github.com/erda-project/erda/modules/pipeline/providers/edgepipeline_register"
+	"github.com/erda-project/erda/modules/pipeline/providers/edgereporter/db"
+	"github.com/erda-project/erda/modules/pipeline/providers/leaderworker"
+)
+
+type config struct {
+	CompensatorDuration time.Duration `file:"COMPENSATOR_DURATION" env:"COMPENSATOR_DURATION" default:"2h"`
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply modules/pipeline/providers/edgereporter/provider.go:34]: done

File: modules/pipeline/providers/edgereporter/interface.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package edgereporter
+
+type Interface interface {
+	AddOneTaskReporter(taskID uint64)
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply modules/pipeline/providers/edgereporter/interface.go:18]: done

File: modules/pipeline/dbclient/op_pipeline_task.go [hunk 4]
```diff
@@ -271,3 +271,11 @@ func (client *Client) CleanPipelineTaskResult(id uint64, ops ...SessionOption) e
 	}
 	return nil
 }
+
+func (client *Client) MultiCreatePipelineTasks(pts []spec.PipelineTask, ops ...SessionOption) (err error) {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply modules/pipeline/dbclient/op_pipeline_task.go:275]: done

File: modules/pipeline/dbclient/op_pipeline_stage.go [hunk 5]
```diff
@@ -108,3 +108,11 @@ func (client *Client) DeletePipelineStagesByPipelineID(pipelineID uint64, ops ..
 		return err
 	}, 3, time.Second)
 }
+
+func (client *Client) MultiCreatePipelineStages(stages []spec.PipelineStage, ops ...SessionOption) (err error) {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply modules/pipeline/dbclient/op_pipeline_stage.go:112]: done

File: apistructs/pipeline.go [hunk 6]
```diff
@@ -701,9 +701,15 @@ type PipelineCallbackResponse struct {
 type PipelineCallbackType string
 
 var (
-	PipelineCallbackTypeOfAction PipelineCallbackType = "ACTION"
+	PipelineCallbackTypeOfAction   PipelineCallbackType = "ACTION"
+	PipelineCallbackTypeOfTask     PipelineCallbackType = "TASK"
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply apistructs/pipeline.go:705]: done

File: modules/pipeline/dbclient/op_pipeline_stage.go [hunk 7]
```diff
@@ -108,3 +108,11 @@ func (client *Client) DeletePipelineStagesByPipelineID(pipelineID uint64, ops ..
 		return err
 	}, 3, time.Second)
 }
+
+func (client *Client) BatchCreatePipelineStages(stages []spec.PipelineStage, ops ...SessionOption) (err error) {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	_, err = session.Table("pipeline_stages").InsertMulti(stages)
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline_stage.go:116]: use spec's table name

File: modules/pipeline/dbclient/op_pipeline_task.go [hunk 8]
```diff
@@ -271,3 +271,11 @@ func (client *Client) CleanPipelineTaskResult(id uint64, ops ...SessionOption) e
 	}
 	return nil
 }
+
+func (client *Client) BatchCreatePipelineTasks(pts []spec.PipelineTask, ops ...SessionOption) (err error) {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	_, err = session.Table("pipeline_tasks").InsertMulti(&pts)
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline_task.go:279]: use spec's table name

File: modules/pipeline/providers/edgereporter/provider.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/edgereporter/provider.go:34]: nothing changed

File: apistructs/pipeline.go [hunk 9]
```diff
@@ -701,9 +701,15 @@ type PipelineCallbackResponse struct {
 type PipelineCallbackType string
 
 var (
-	PipelineCallbackTypeOfAction PipelineCallbackType = "ACTION"
+	PipelineCallbackTypeOfAction             PipelineCallbackType = "ACTION"
+	PipelineCallbackTypeOfEdgeTaskReport     PipelineCallbackType = "TASK"
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply apistructs/pipeline.go:705]: done

File: modules/pipeline/dbclient/op_pipeline_task.go (hunk 8 above)
Comment (littlejiancc) [reply modules/pipeline/dbclient/op_pipeline_task.go:279]: done

File: modules/pipeline/dbclient/op_pipeline_stage.go (hunk 7 above)
Comment (littlejiancc) [reply modules/pipeline/dbclient/op_pipeline_stage.go:116]: done

**Review by sfwn — APPROVED** [review]

**Pull request merged**